### DISCUSSION TITLE
Disabling automerge because PR was merged previously

### DISCUSSION THREAD

**Discussion by adam-moss** [renovatebot/renovate#27945] [Request Help]
### What would you like help with?

Other

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab with Renovate official docker image at latest

### Please tell us more about your question or problem

I've been seeing an increasing number of updates I'm expecting to automerge result in a merge request being raised instead.

Running renovate with a log level of debug gives the following message:

> Disabling automerge because PR was merged previously

I understand this was done previously, quite rightly, to avoid the scenario where something was automerged and subsequently rolled back to avoid undesirable side affects.

What I'm not clear on though is why it is happening in my scenario:

Merge request just created:

| Package | Type | Update | Change | OpenSSF |
|---|---|---|---|---|
| cgr.dev/chainguard/python | final | digest | `9ee3605` -> `9098455` |  |
| cgr.dev/chainguard/python | stage | digest | `9ee3605` -> `9098455` |  |
| [hypothesis](https://hypothesis.works) ([source](https://github.com/HypothesisWorks/hypothesis/tree/HEAD/hypothesis-python), [changelog](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/docs/changes.rst)) | dev | patch | `6.99.5` -> `6.99.6` | [![OpenSSF Scorecard](https://api.securityscorecards.dev/projects/github.com/HypothesisWorks/hypothesis/badge)](https://securityscorecards.dev/viewer/?uri=github.com/HypothesisWorks/hypothesis) |

Merge request it has been matched against:

| Package | Type | Update | Change | OpenSSF |
|---|---|---|---|---|
| cgr.dev/chainguard/python | final | digest | `202e5d1` -> `9ee3605` |  |
| cgr.dev/chainguard/python | stage | digest | `202e5d1` -> `9ee3605` |  |
| [hypothesis](https://hypothesis.works) ([source](https://github.com/HypothesisWorks/hypothesis/tree/HEAD/hypothesis-python), [changelog](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/docs/changes.rst)) | dev | patch | `6.99.2` -> `6.99.5` | [![OpenSSF Scorecard](https://api.securityscorecards.dev/projects/github.com/HypothesisWorks/hypothesis/badge)](https://securityscorecards.dev/viewer/?uri=github.com/HypothesisWorks/hypothesis) |

So whilst they are the same dependencies, they're different versions.  And shouldv'e been merged by branch rather than with an MR anyway 🤷 .

The associated repository config file is simple:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [],
  "packageRules": [
    {
      "groupName": "churn-updates",
      "addLabels": [
        "Automatic Merge"
      ],
      "matchUpdateTypes": [
        "digest",
        "minor",
        "patch"
      ],
      "automerge": true,
      "automergeType": "branch",
      "platformAutomerge": true,
      "gitLabIgnoreApprovals": true,
      "commitMessageAction": "",
      "commitMessageExtra": ""
    }
  ]
}
```

What would cause this behaviour and how can I "reset" it?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
Are these "immortal" PRs?

**Reply by adam-moss** [reply]
Nothing I can see that suggests they are and the only ones we intentionally `recreateWhen: always` is for dependency pinning

**Reply by rarkins** [reply]
What is an example PR title?

**Reply by adam-moss** [reply]
Title is always the same, in this case `chore: churn-updates`

**Reply by rarkins** [reply]
That sounds like an immortal PR

**Reply by adam-moss** [reply]
Ok, I *think* I get what you're driving at, this is a cache-key issue, right?

The `config.js` essentially has:

```javascript
    ...
    branchPrefix: "renovate/chore",
    semanticCommits: "enabled",
    semanticCommitType: "chore",
    semanticCommitScope: "",
    ...
```

When combined with the above repo config that is resulting in consistent:

- branch name of `renovate/chore/churn-updates`
- MR title of `chore: churn-updates`

Although the MRs themselves don't have the usual

> 👻 Immortal: This PR will be recreated if closed unmerged.

So yes, it does indeed sound like this is what may be causing it.   But... what I'm not clear on is _why_.  Per the config above the `automergeType` is `branch` so an MR should never have been created to begin with 🤔 

Nevertheless it sounds like I need to either:

1. not group, and instead apply the `automerge` directives against the higher level `digest`, `minor`, and `patch` update types, or
2. keep the group and make the branch name more dynamic, e.g. injecting a random string or the execution timestamp or similar

Is my understanding here correct?  If some I presume (1) is the "safer" of the two approaches?

Thanks!

**Reply by rarkins** [reply]
A PR which has a generic title like "chore: churn-updates" should normally be classified as immortal and then _not_ have limits about automerging multiple times. The automerge protection should only apply for mortal PRs which can be identified as duplicates.

Possible ways forward:
1. Try to diagnose from logs why this MR is not immortal
2. Create a minimal reproduction of an immortal MR proving that there's a bug and what I wrote above is false

**Reply by adam-moss** [reply]
Will look at creating a repro 👍 

For now I've simply updated the repo config to sidestep it, although it is slightly less convenient as it requires more reno runs to be performed:

```yaml
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "packageRules": [
    {
      "automerge": true,
      "automergeType": "branch",
      "matchCurrentVersion": "!/^0/",
      "matchUpdateTypes": ["digest", "minor", "patch", "pin"]
    }
  ],
}
```
